File: code_snippets/tree_bst.js
// assume Node structure has .right, .left, and .data methods.
// .right or .left called on a leaf will return null

function isBST(root, maxSeen, minSeen) {
  // base case
  if(root === null) {
    return true;
  }

  // checks if the function has traversed nodes
  // and that the root's value is smaller than the max value seen
  if (maxSeen !== undefined && maxSeen < root.data) {
    return false;
  }

  if (minSeen !== undefined && minSeen > root.data) {
    return false;
  }

  const leftResult = isBST(root.left, root.data - 1, minSeen);
  const rightResult = isBST(root.right, maxSeen, root.data + 1);

  return(leftResult && rightResult);
}

File: code_snippets/tree_bst.py
def is_bst(root, max_seen = None, min_seen = None):
    if root == None:
        return True
    if (max_seen and max_seen < root.data):
        return False
    if (min_seen and min_seen > root.data):
        return False
    left_result = is_bst(root.left, root.data - 1, min_seen)
    right_result = is_bst(root.left, max_seen, root.data + 1)
    return(left_result and right_result)
